File: packages/core/src/services/ExportService.ts
import { noteService } from '../services';
import type { Note } from '../models';

/**
 * Export notes to various formats
 */
export class ExportService {
  /**
   * Export a single note to markdown
   */
  noteToMarkdown(note: Note): string {
    const userTags = note.tags
      .filter(t => !t.startsWith('backlink:') && !t.startsWith('outlink:'))
      .map(t => `#${t}`)
      .join(' ');

    const frontmatter = [
      '---',
      `title: "${note.content.title}"`,
      `created: ${new Date(note.created_at).toISOString()}`,
      `updated: ${new Date(note.updated_at).toISOString()}`,
      userTags ? `tags: [${note.tags.filter(t => !t.startsWith('backlink:') && !t.startsWith('outlink:')).map(t => `"${t}"`).join(', ')}]` : '',
      '---',
      '',
    ].filter(Boolean).join('\n');

    return frontmatter + note.content.body;
  }

  /**
   * Export all notes to a real ZIP file using JSZip
   */
  async exportToZip(): Promise<Blob> {
    const JSZip = (await import('jszip')).default;
    const zip = new JSZip();
    const notes = await noteService.list();

    for (const note of notes) {
      const safeName = note.content.title
        .replace(/[/\\?%*:|"<>]/g, '-')
        .slice(0, 50) || 'untitled';
      
      zip.file(`${safeName}.md`, this.noteToMarkdown(note));
    }

    return await zip.generateAsync({ type: 'blob' });
  }

  /**
   * Download a file in the browser
   */
  downloadFile(content: Blob | string, filename: string) {
    const blob = typeof content === 'string' 
      ? new Blob([content], { type: 'text/markdown' })
      : content;
    
    const url = URL.createObjectURL(blob);
    const a = document.createElement('a');
    a.href = url;
    a.download = filename;
    document.body.appendChild(a);
    a.click();
    document.body.removeChild(a);
    URL.revokeObjectURL(url);
  }

  /**
   * Export single note and trigger download
   */
  async downloadNote(note: Note) {
    const content = this.noteToMarkdown(note);
    const safeName = note.content.title.replace(/[^a-z0-9]/gi, '_').slice(0, 50);
    this.downloadFile(content, `${safeName}.md`);
  }

  /**
   * Export all notes and trigger download
   */
  async downloadAll() {
    const blob = await this.exportToZip();
    const date = new Date().toISOString().split('T')[0];
    this.downloadFile(blob, `relix-export-${date}.zip`);
  }
}

export const exportService = new ExportService();

File: frontend/lib/services/export_service.dart
import 'dart:convert';
import 'dart:typed_data';

import 'package:archive/archive.dart';
import 'package:intl/intl.dart';
import 'package:share_plus/share_plus.dart';

import '../models.dart';

class ExportService {
  String noteToMarkdown(NoteEntry entry) {
    if (entry.type != 'note') return 'Not a note entry.';
    final note = entry.content as NoteContent;

    final userTags = entry.tags
        .where((t) => !t.startsWith('backlink:') && !t.startsWith('outlink:'))
        .toList();

    final createdAt = DateTime.fromMillisecondsSinceEpoch(
      entry.createdAt * 1000,
    );
    final updatedAt = DateTime.fromMillisecondsSinceEpoch(
      entry.updatedAt * 1000,
    );
    final isoFormat = DateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'");

    final frontmatter = [
      '---',
      'title: "${note.title}"',
      'created: ${isoFormat.format(createdAt.toUtc())}',
      'updated: ${isoFormat.format(updatedAt.toUtc())}',
      if (userTags.isNotEmpty)
        'tags: [${userTags.map((t) => '"$t"').join(', ')}]',
      '---',
      '',
    ].join('\n');

    return frontmatter + note.body;
  }

  Future<Uint8List> exportToZip(List<NoteEntry> notes) async {
    final archive = Archive();

    for (final note in notes) {
      if (note.type != 'note') continue;
      final content = note.content as NoteContent;

      final safeName = content.title
          .replaceAll(RegExp(r'[/\\?%*:|"<> \.]'), '-')
          .substring(0, content.title.length > 50 ? 50 : content.title.length);

      final fileName = '${safeName.isEmpty ? 'note' : safeName}-${note.id}.md';
      final bytes = Uint8List.fromList(utf8.encode(noteToMarkdown(note)));
      archive.addFile(ArchiveFile(fileName, bytes.length, bytes));
    }

    final encoder = ZipEncoder();
    return Uint8List.fromList(encoder.encode(archive));
  }

  Future<void> shareNote(NoteEntry note) async {
    final content = noteToMarkdown(note);
    final noteContent = note.content as NoteContent;
    final safeName = noteContent.title
        .replaceAll(RegExp(r'[^a-z0-9]'), '_')
        .toLowerCase();

    await SharePlus.instance.share(
      ShareParams(
        files: [
          XFile.fromData(
            Uint8List.fromList(utf8.encode(content)),
            name: '${safeName.isEmpty ? 'relix-note' : safeName}-${note.id}.md',
            mimeType: 'text/markdown',
          ),
        ],
        text: 'Relix Export: ${noteContent.title}',
      ),
    );
  }

  Future<void> shareAll(List<NoteEntry> notes) async {
    final zipBuffer = await exportToZip(notes);
    final date = DateFormat('yyyy-MM-dd').format(DateTime.now());
    await SharePlus.instance.share(
      ShareParams(
        files: [
          XFile.fromData(
            zipBuffer,
            name: 'relix-export-$date.zip',
            mimeType: 'application/zip',
          ),
        ],
        text: 'Relix Backup ($date)',
      ),
    );
  }
}
